synapse/storage/databases/main/events_bg_updates.py

```diff
@@ -1023,10 +1034,10 @@ def _calculate_chain_cover_txn(
         PersistEventsStore._add_chain_cover_index(
             txn,
             self.db_pool,
-            self.event_chain_id_gen,
+            self.event_chain_id_gen,  # type: ignore[attr-defined]
             event_to_room_id,
             event_to_types,
-            event_to_auth_chain,
+            cast(Dict[str, Sequence[str]], event_to_auth_chain),
```

Comment:
If not cast here:

```sh
synapse/storage/databases/main/events_bg_updates.py:1040: error: Argument 6 to "_add_chain_cover_index" of "PersistEventsStore" has incompatible type "Dict[str, List[str]]"; expected "Dict[str, Sequence[str]]"  [arg-type]
synapse/storage/databases/main/events_bg_updates.py:1040: note: "Dict" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
synapse/storage/databases/main/events_bg_updates.py:1040: note: Consider using "Mapping" instead, which is covariant in the value type
```

Is there a better solution?

Reply:
I'm unsure if there's a better solution, this seems like it might be fine? In this method the sequence is mutable, while in `_add_chain_cover_index` it doesn't ever need it to be mutable so we're telling mypy it can treat the second argument as non-mutable.

Reply:
This part of Python's type system is always a pain.
There's no nice solution here, as much as I always try to see one.
To get around it without having to cheat the system, you'd have to be able to 'consume' a `Dict[str, List[str]]` and turn it into `Dict[str, Sequence[str]]` such that you couldn't ever re-use the `Dict[str, List[str]]` (since it might contain e.g. tuples).

You could probably do that by cloning it, since the clone and the old one wouldn't be the same thing, but that seems like a waste of time/memory when the dict has already been built.

Realistically, you're probably better off leaving this `cast` in, as much as that makes me sad. With any luck there are some tests that check this part of the code that will ensure the types are right... :-)